zebra-network/src/meta_addr.rs

```diff
@@ -17,55 +19,314 @@ use zebra_chain::serialization::{
 
 use crate::protocol::{external::MAX_PROTOCOL_MESSAGE_LEN, types::PeerServices};
 
+use MetaAddrChange::*;
 use PeerAddrState::*;
 
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest::option;
 #[cfg(any(test, feature = "proptest-impl"))]
 use proptest_derive::Arbitrary;
+
 #[cfg(any(test, feature = "proptest-impl"))]
-mod arbitrary;
+use zebra_chain::primitives::arbitrary::{datetime_full, datetime_u32};
 
 #[cfg(test)]
 mod tests;
 
 /// Peer connection state, based on our interactions with the peer.
 ///
-/// Zebra also tracks how recently a peer has sent us messages, and derives peer
-/// liveness based on the current time. This derived state is tracked using
-/// [`AddressBook::maybe_connected_peers`] and
-/// [`AddressBook::reconnection_peers`].
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+/// Zebra also tracks how recently we've had different kinds of interactions with
+/// each peer, and derives peer usage based on the current time. This derived state
+/// is tracked using [`AddressBook::recently_used_peers`] and
+/// [`AddressBook::candidate_peers`].
+///
+/// To avoid depending on untrusted or default data, Zebra tracks the required
+/// and optional data in each state. State updates are applied using
+/// [`MetaAddrChange`]s.
+///
+/// See the [`CandidateSet`] for a detailed peer state diagram.
+#[derive(Copy, Clone, Debug)]
 #[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
 pub enum PeerAddrState {
-    /// The peer has sent us a valid message.
+    /// The peer's address has just been fetched from a DNS seeder.
     ///
-    /// Peers remain in this state, even if they stop responding to requests.
-    /// (Peer liveness is derived from the `last_seen` timestamp, and the current
-    /// time.)
-    Responded,
+    /// Zebra attempts to connect to all DNS seeder peers on startup. So the
+    /// initial peer connector updates all DNS seeder peers to `AttemptPending`.
+    ///
+    /// If any DNS seeder addresses remain, they are attempted after disconnected
+    /// `Responded` peers.
+    NeverAttemptedDnsSeeder,
 
-    /// The peer's address has just been fetched from a DNS seeder, or via peer
-    /// gossip, but we haven't attempted to connect to it yet.
-    NeverAttemptedGossiped,
+    /// The peer's address has just been fetched via peer gossip, but we haven't
+    /// attempted to connect to it yet.
+    ///
+    /// Gossiped addresses are attempted after DNS seeder addresses.
+    NeverAttemptedGossiped {
+        /// The time that another node claims to have connected to this peer.
+        /// See `get_untrusted_last_seen` for details.
+        #[cfg_attr(
+            any(test, feature = "proptest-impl"),
+            proptest(strategy = "datetime_u32()")
+        )]
+        untrusted_last_seen: DateTime<Utc>,
```

Comment:
Open a follow-up ticket to:
- make wrapper types for 32-bit, 64-bit, and non-serialized times
- make arbitrary and serialization impls for those wrappers
- delete the unused strategies

Reply:
#2171